atc/lidar/scanner.go

```diff
@@ -1,28 +1,63 @@
 package lidar
 
 import (
+	"code.cloudfoundry.org/lager/lagerctx"
 	"context"
+	"encoding/json"
 	"strconv"
 	"sync"
 
-	"code.cloudfoundry.org/lager/lagerctx"
+	"github.com/concourse/concourse/atc/component"
 	"github.com/concourse/concourse/atc/db"
 	"github.com/concourse/concourse/atc/metric"
 	"github.com/concourse/concourse/atc/util"
 	"github.com/concourse/concourse/tracing"
 )
 
-func NewScanner(checkFactory db.CheckFactory) *scanner {
+func NewScanner(checkFactory db.CheckFactory, chunks int) *scanner {
 	return &scanner{
 		checkFactory: checkFactory,
+		chunks:       chunks,
+	}
+}
+
+type scannerRunResult struct {
+	LastScannedResourceId     int `json:"last_scanned_resource_id"`
+	LastScannedResourceTypeId int `json:"last_scanned_resource_type_id"`
+}
+
+func parseLastScanResult(result string) scannerRunResult {
+	sr := scannerRunResult{
+		LastScannedResourceId:     -1,
+		LastScannedResourceTypeId: -1,
 	}
+	if result == "" {
+		return sr
+	}
+
+	// Ignore json decode error. If the field is never set or crashes, it will
+	// be auto fixed in next run.
+	json.Unmarshal([]byte(result), &sr)
+	return sr
+}
+
+func (r scannerRunResult) String() string {
+	b, err := json.Marshal(r)
+	if err != nil {
+		return ""
+	}
+	return string(b)
 }
 
 type scanner struct {
 	checkFactory db.CheckFactory
+	chunks       int
 }
 
-func (s *scanner) Run(ctx context.Context) error {
+// Run processes 1/chunks of resources and resource types in each period, which
+// helps distribute Lidar checks to different ATCs. The last scanned resource id
+// and resource type id are stored in the component last run result.
```

Comment:
I don't think I'm fully grasping the motivation for the chunking/batching in Lidar. AFAIK, the main reason is to help prevent all checks from running on the same ATC, right?

Suppose there are 60 resources (numbered 1-60), and 6 chunks. Lidar will first run and create checks for resources 1-10. 10 seconds later, it'll run again and create checks for 11-20. And so on until all the resources are checked.

But is this necessary? Because of the rate limiting, the checks will be run evenly over the 1m interval - and since Lidar (or specifically the `CheckFactory`) doesn't create a check if the resource's interval hasn't elapsed, shouldn't we only end up queueing only ~1/6th of the checks in each `Scanner.Run` call anyway?

Reply:
I added chunks after I ran some load tests.

Before in-memory check, Lidar will insert check builds into the db, then trackers on different ATCs will share check builds. With in-memory, checks will run on the same ATC where Lidar runs, which will cause a spike of goroutines on an ATC, especially on a large deployments that have a lot of resources. Chunk helps distribute checks among multiple ATCs. 

I don't understand how rate limit will end up queuing ~1/6 checks. If all checks succeed quickly, all 60 resources will finishes in the first 10 seconds, then rest 50 seconds will not run any check. Then in next minute's first 10 seconds, 60 checks will run again.

Reply:
The default rate limiting behaviour should prevent all of the checks from running within the same 10 second span. If there are 60 resources, with the default resource check interval of 1m, then the default rate limit is 1 check/second. So, in 10 seconds, only 10 checks should be able to be run.

I notice you had set `CONCOURSE_MAX_CHECKS_PER_SECOND: -1` in the `docker-compose.yml`, and later commented it out - did you have that configuration in your load testing? As that would bypass the rate limiting behaviour I'm describing.

Reply:
Actually, I *think* I see the problem now. Assuming there's just 1 ATC for now:

* After the first lidar interval, 60 check goroutines have been created
* By the start of the second lidar interval, 10 of those goroutines have exited since the check ran - the rest are rate limited
* The second lidar interval skips over the first 10 resources since they get filtered out in https://github.com/concourse/concourse/blob/ad613218411dbbbd80cf325f45a4bc90b215edaf/atc/db/check_factory.go#L125
  * However, it creates **50 more check goroutines** as the current 50 waiting resources haven't yet updated their `LastCheckStartTime`

So I think the problem may be that there's nothing preventing us from creating a new in-memory check for a resource if another is already queued up.

What if instead of chunking/batching lidar, we keep track of the resource IDs that already have checks in memory and only create checks if the resource ID isn't in the set. When a check is created, we add its resource ID to the set. When the check finishes, we remove it.

If there are multiple ATCs, each one may have a goroutine for each resource, but at least there can only be one goroutine per resource. This should incur some memory overhead, but it shouldn't be too crazy hopefully. Each goroutine consumes at least 2KiB in stack. Assuming 10x that for the full memory usage of a check (20KiB/check), 10,000 checks should only "cost" 20MiB. Making a lot of assumptions here, this will need to be measured (and I could be way off).

Reply:
Let me set chunks to 1 in today's test which will be same as the original logic without chunks and see what happens. If rate limit is good enough, I'll remove chunk related code later.

Reply:
I guess my point is that with rate limiting (or with anything that would prevent checks from running right away/taking a long time to run), and without a way to prevent queueing multiple checks for the same resource on a single ATC, the ATCs will end up using a lot more memory than is required (especially if the default check interval is long). So I still think some kind of in-memory set of "active" resource checks would be useful

Reply:
This is a good point, which makes me recall `inMemoryCheckBuild.AcquireTrackingLock()`. Initially I made this function just returned a no-op lock. I just changed it to acquire a lock on resource/resourceType id, so that, when a check is slow, other in-mem check builds on the same resource/resourceType will quickly terminate.

WRT chunks, I think it's still helpful for large deployment. Say there are 600 resources. Lidar scan starts on ATC1, and in the first 10 seconds, 60 resource builds are started, rest are waiting on rate limit; Then next Lidar scan starts on ATC2, the first 10 seconds will wait on lock (or abort with this PR's logic), rest will wait on rate limit. In the meantime, The first Lidar scan on ATC1 will process 60 more resources. With this logic, all resource will only be processed on ATC1 for the minute.

With chunks, 60 resources will be processed on ATC1 in the first interval, and last processed id is stored in db; next scan on ATC2 will processes the following 60 resources, and so on ... which helps distribute Lidar checks to different ATCs.

Reply:
The new `inMemoryCheckBuild.AcquireTrackingLock()` behaviour simplifies things somewhat since it means there can only be one running check for a given resource/resource type across *all* ATCs.

I kind of see what you mean about chunking, but I think it should balance out over time. Let's suppose there are 600 resources and no chunking.

* The first pass will result in ATC1 starting all 600 checks. 100 are able to run in the 10s interval, and the rest are waiting on the rate limit.
* ATC2 will then try to create checks for the last 500 resources, but will fail to do so (since ATC1 holds the locks). It skips the first 100 because the interval has not elapsed. This cycle repeats for 4 more cycles, with the number of checks that can't be created due to the lock being held decreasing by 100 each time.
* ATC2 starts the 6th lidar scan - it will try to create checks for only the first 100 resources (the rest haven't completed their check interval). At this point, these are the only checks that are able to run.
* ATC1 starts the 7th lidar scan, and creates checks for only the next 100 resources (again, the rest haven't completed their check interval)
* ...and so on

I may be missing something (and do let me know if I am), but with this assessment, it seems like checks should be relatively spaced out even without chunking, with the exception of the very first batch. Note that this only works because of the modified `inMemoryCheckBuild.AcquireTrackingLock()` behaviour.

By the way, I'm mostly just trying to work out whether we actually need this chunking, because having two mechanisms for distributing checks over time (namely, lidar chunking and rate limiting) feels odd (and difficult to troubleshoot). It may be that we can actually get rid of rate limiting in favour of chunking at the check creation stage, though I'm a bit wary to do that without better understanding the consequences. Also tagging @clarafu who has a bit more context here